### PR TITLE
[Arc] Add ArcCanonicalizer pass

### PR DESCRIPTION
**Pull request by maerhart**
Add a pass that allows us to implement canonicalization patterns that require a symbol cache to work efficiently (as regular canonicalization patterns don't get one provided). But also allows us to add canonicalizations to operations from other dialects that are simulation specific and thus not desirable to be added to that other dialect directly.

### CODE REVIEW

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+private:
+  DenseMap<Operation *, SetVector<Operation *>> userMap;
+};
+
+//===----------------------------------------------------------------------===//
+// Canonicalization patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// A rewrite patten that has access to a symbol cache to access and modify the
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:104]:
```suggestion
/// A rewrite pattern that has access to a symbol cache to access and modify the
```

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+
+LogicalResult MemWritePortEnableAndMaskCanonicalizer::matchAndRewrite(
+    MemoryWritePortOp op, PatternRewriter &rewriter) const {
+  auto defOp = cast<DefineOp>(symbolCache.getDefinition(op.getArcAttr()));
+  APInt enable;
+
+  if (op.getEnable() &&
+      mlir::matchPattern(
+          defOp.getBodyBlock().getTerminator()->getOperand(op.getEnableIdx()),
+          mlir::m_ConstantInt(&enable))) {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:178]: That's a very neat way to look through the symbol and match against something in the arc definition!

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+            rewriter.getFunctionType(defOp.getArgumentTypes(), newResultTypes));
+      });
+
+      // Update symbol cache
+      symbolCache.addDefinition(defOp.getNameAttr(), defOp);
+      symbolCache.addDefinition(newDefOp.getNameAttr(), newDefOp);
+      symbolCache.addUser(defOp, callOp);
+      for (auto *user : users)
+        symbolCache.addUser(user == op ? defOp : newDefOp, user);
+      return success();
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:232]: Should this pattern have a cache somewhere that keeps track of the defOp-minus-enable for each defOp, such that you only create a def without enable for the first memory op, and subsequent ops would just use the cached def without enable?

Probably not a big thing in practice though :grin:

**Review by fabianschuiki — APPROVED** [review]
Very nice! LGTM

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp (hunk 3 above)
Comment (maerhart) [reply lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp]: Excellent point! I've added that.

**Pull request merged**